### DISCUSSION TITLE
Not correct modulated fsk

### DISCUSSION THREAD

**Discussion by A-SOM** [jgromes/RadioLib#751] [Q&A]
Module: SX1262

I have code:
```c++
#include <RadioLib.h>
SX1262 radio = new Module(RADIOLIB_BUILTIN_MODULE);

void setup() {
  Serial.begin(115200);
  Serial.print(F("[SX1262] Initializing ... "));
  int state = radio.beginFSK();
  if (state != RADIOLIB_ERR_NONE) {
    Serial.println(F("failed!"));
    return;
  }
  state = radio.setFrequency(868.0);
  state = radio.setBandwidth(187.2);
  state = radio.setBandwidth(250.0);
  state = radio.setOutputPower(10);
  state = radio.setBitRate(10.0);
  state = radio.setFrequencyDeviation(4.5);
  state = radio.setCRC(0);
  state = radio.setCurrentLimit(100.0);
  state = radio.setPreambleLength(16);
  state = radio.disableAddressFiltering();
  state = radio.setEncoding(RADIOLIB_ENCODING_NRZ);
  Serial.println(F("success!"));
  uint8_t syncWord[2] = {0x55, 0x55};
  state = radio.setSyncBits(syncWord, 8);
}
void loop() {
  String bitString = "00010110111010100000110110110111100000000000001111000000011001000000111100001111001000001111100011010101111100110001001001110011010100100100100000001000000010000000100000001000000010000000100000001000000010000000100000001000000010000000";
  byte data[bitString.length()/8];
  
  for (int i = 0; i < bitString.length(); i+=8) {
    data[i/8] = strtol(bitString.substring(i, i+8).c_str(), 0, 2);
  }
  Serial.println(F("SEND DATA!"));
  int state = radio.transmit(data, sizeof(data));

  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Done!"));
  } else {
    Serial.print(F(Error Send: "));
    Serial.println(state);
  }
  delay(1000);
}
```

I want recvie: 555516ea0db78003c0640f0f20f8d5f31273524808080808080808080808080 (hex)
but i take: f55551d16ea0db78003c0640f0f20f8d5f3127352480808080808080808080808e

how i can remove at start "f" and after preambule "1d" and in ending take corect symvol?

**Comment by jgromes** [comment]
Couple of observations:

1. You're attempting to set sync word to 0x5555, but only setting eight bits. Is that intentional? Also, why are you setting this as the sync word? It's an alternating pattern of 0s and 1s, which is usually used as a preamble. Sync word should be different from preamble.

2. You seem to have 236 bits in your "binary string". That's not a number that can be divided by 8, so you are probably miscounting something. The SX126x can only transmit complete bytes when sending a packet. I don't know what exactly you are trying to do, but surely saving the data as a simple array would be orders of magnitude faster, easier to understand and debug: `byte data[] = {0x16, 0xea, 0x0d ... `

3. You have not changed the default packet mode, which is variable packet length. The extra 0x1D is probably the packet length.

> but i take: f55551d16ea0db78003c0640f0f20f8d5f3127352480808080808080808080808e

How are you receiving this?

**Comment by A-SOM** [comment]
Thank you answer.
> How are you receiving this?

I use rtl-sdr + Universal Radio Hacer

1. I'm sorry, I'm not experienced in fsk
I just want to repeat the signal that my device makes with lora
It seemed to me that the welcoming speech was the preamble

2. Yes it can be, because i try repeat signal from URH 

3. How i can do it?
I want repeat signal from: http://ww1.microchip.com/downloads/en/devicedoc/70590c.pdf
and i almost do it..
my original signal from URH: 010101010101010100010110111010100000110110110111100000000000001111000000011001000000111100001111001000001111100011010101111100110001001001110011010100100100100000001000000010000000100000001000000010000000100000001000000010000000100000001000000010000000
hex (555516ea0db78003c0640f0f20f8d5f31273524808080808080808080808080)

**Comment by jgromes** [comment]
If the data is coming directly from URH, then it is likely the extra 0xf and 0xe at the beginning and end are just the amplifier of the SX126x ramping up and down. The device you are sending this to won't actually see it, because in the beginning it searches for the preamble and at the end it either has a fixed packet length that it expects, or has the packet length encoded somewhere within the data.

So the only two things you have to do are:

1. Set fixed packet mode by calling `radio.fixedPacketLengthMode(len);` and setting your data length in bytes as the `len` argument.
2. Make sure you are not trying to send number of bits that cannot be divided by 8.

**Comment by A-SOM** [comment]
I almost resolve all.
but exta 0xf it not good for me. how i can remove it?
055550416ea0db78003c0640f0f20f8d5f31273524808080808080808080808080
^

**Comment by jgromes** [comment]
I don't see an extra 0xf in what you posted. I would suggest to stop using that binary string, any instead focusing on reverse engineering the packet itself - URH has lots of features to help with that. Is there some synchronization word? What is the packet format? etc. If you do that, you will be able to better understand what is going wrong instead of just guessing.

Also - this doesn't seem like an issue in the lbirary, so I'm converting it to discussion.

**Comment by A-SOM** [comment]
I want remove first symvol at start "0"
0555516ea0db78003c0640f0f20f8d5f31273524808080808080808080808080
0 < - wtf ?
5555 - standart fsk preambule
16ea0db78003c0640f0f20f8d5f31273524808080808080808080808080 <- it my data packet

**Comment by A-SOM** [comment]
my config

```c++
#include <RadioLib.h>
SX1262 radio = new Module(RADIOLIB_BUILTIN_MODULE);

void setup() {
  Serial.begin(115200);
  Serial.print(F("[SX1262] Initializing ... "));
  int state = radio.beginFSK();
  if (state != RADIOLIB_ERR_NONE) {
    Serial.println(F("failed!"));
    return;
  }
  state = radio.setFrequency(867.9785);
    state = radio.setBandwidth(250.0);
  state = radio.setOutputPower(10);
  state = radio.setBitRate(10.0);
  //100%
  state = radio.setFrequencyDeviation(240);
  state = radio.setCRC(0);
  state = radio.disableAddressFiltering();
  state = radio.fixedPacketLengthMode(31);
  // state = radio.setWhitening(false);
  //### work only with it
  state = radio.setDio2AsRfSwitch(false);
  state = radio.implicitHeader(0);
  // state = radio.setDataShaping(RADIOLIB_SHAPING_NONE);
  state = radio.setDataShaping(RADIOLIB_SHAPING_0_3);
  // state = radio.setDataShaping(RADIOLIB_SHAPING_0_5);
  // state = radio.setDataShaping(RADIOLIB_SHAPING_0_7);
  // state = radio.setDataShaping(RADIOLIB_SHAPING_1_0);
  uint8_t syncWord[0] = {};
  state = radio.setSyncBits(syncWord, 0);
}
```

**Reply by jgromes** [reply]
As I wrote above:

> If the data is coming directly from URH, then it is likely the extra 0xf and 0xe at the beginning and end are just the amplifier of the SX126x ramping up and down. The device you are sending this to won't actually see it, because in the beginning it searches for the preamble

The transceiver you are transmitting to will lock on the 0x5555 preamble - that's why it is there. Anything in front of it doesn't matter. Just because you see it in URH doesn't mean it will be received like this.

Additionally, some of the configfuration you are doing is not applicable in FSK mode, which you would have known had you read the documentation, or even checked the return code. Specifically the following functions are only available in LoRa mode:

```c++
radio.setBandwidth(250.0);
radio.implicitHeader(0);
```

**Reply by A-SOM** [reply]
Yes, I know these features are not available. This is a bookmark for the future. How setFrequencyDeviation 240 (I modified source for it, and it work)
I received in spectogram from MRF49XA  it:
https://prnt.sc/4lPpTWxJTvc6
My signal now, it from SX1262 . I can't repeat it signal for my task :(
https://prnt.sc/McHH3OLdh5GU
when i disable DataShaping, i take bad data in URH :(

**Comment by A-SOM** [comment]
@jgromes hi.
i und what i do wrong.
SX1262 work in gfsk as default. how i can enable pure fsk?

**Reply by jgromes** [reply]
SX1268 uses unshaped FSK by default - you have enabled shaping manually by `radio.setDataShaping(RADIOLIB_SHAPING_0_3);`

Either don't call this method at all, or call it with the parameter `RADIOLIB_SHAPING_NONE`
